tensorflow_hub/hub2graph.py

```diff
@@ -0,0 +1,143 @@
+"""
+Fetch a module from Hub and write it as GraphDef binary Protocol Buffers file
+and also optionally as Tensorboard logs. The former is useful for
+running inference in other languages with lagging API support, e.g. Go.
+The latter allows to visualize the model in Tensorboard.
+
+Usage:
+
+```
+python3 /path/to/hub2graph.py -m <TF Hub path or URL> -o output_graph_def.pb
+```
+"""
+
+import argparse
+import logging
+import sys
+import os
+
+import tensorflow as tf
+from tensorflow.python.framework import graph_util, graph_io
+from tensorflow.python.summary import summary
+import tensorflow_hub as tfhub
+
+
+def setup():
+    """
+    Parse command line arguments and setup logging.
+    """
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument("-m", "--hub", required=True,
+                        help="Tensorflow Hub module URL.")
+    parser.add_argument("-o", "--output", required=True,
+                        help="Output graph definition.")
+    parser.add_argument("-t", "--tensorboard",
+                        help="Tensorboard output log directory.")
+    parser.add_argument("--disable-optimization", action="store_true",
+                        help="Do not exclude nodes unrelated to inference.")
+    parser.add_argument("-v", "--log-level", choices=logging._nameToLevel,
+                        default="WARN")
+    args = parser.parse_args()
+    level = logging._nameToLevel[args.log_level]
+    logging.basicConfig(level=level)
+    os.environ["TF_CPP_MIN_LOG_LEVEL"] = str(level // 10 - 2)
+    return args
+
+
+def tensor_name(t):
+    """
+    Prepends "module" to tensor name and removes ":0" in the end
+    """
+    return "module/" + t.name.rsplit(":", 1)[0]
+
+
+def print_signature(sig, name):
+    """
+    Prints the Markdown table with inputs or outputs of a Hub module.
+    """
+    label_size = max([len(k) for k in sig])
+    name_size = max([len(tensor_name(v)) for _, v in sig.items()])
+    print("\n%s\n%s\n\n| label%s | tensor%s |\n|:------%s|:-------%s|" % (
+        name,
+        "=" * len(name),
+        " " * (label_size - 5) if label_size > 5 else "",
+        " " * (name_size - 6),
+        "-" * (label_size - 5) if label_size > 5 else "",
+        "-" * (name_size - 6)
+    ))
+    for k, v in sorted(sig.items()):
+        print("| %s%s | %s%s |" % (
+            k, " " * (label_size - len(k)),
+            tensor_name(v), " " * (name_size - len(tensor_name(v)))
+        ))
+    print()
+
+
+def main():
+    """
+    Performs the conversion from Hub module to GraphDef + TB.
+    """
+    args = setup()
+    log = logging.getLogger("main")
+    with tf.Graph().as_default() as graph:
+        # everything happens inside the single default graph
+        log.info("Importing %s", args.hub)
+        module = tfhub.Module(args.hub)
+        signature_def = module._impl._meta_graph.signature_def.get(
```

Comment:
Could you create tf.placeholders with known names and feed them to module.__call__() instead of depending on internals?

Reply:
I had `__call__` in the initial version of the script  it there are problems with it. It creates a new subgraph and links variables from the initial graph. It takes considerable time (up to 10 seconds). Besides, it gives a nasty long name to the subgraph (totally sane for the intended use case) which is hard to change. It also inserts placeholders before placeholders - working solution, but not the cleanest for Tensorboard.